Use case: Databricks (#2455)
astronomer/docs: learn/use-case-airflow-databricks.md

---

> title: "ELT pipelines with Airflow, Databricks, the Astro Databricks provider, and the Astro Python SDK"

Reviewer: ```suggestion
title: "ELT pipelines with Airflow, the Astro Databricks provider, and the Astro Python SDK"
```
I think "databricks" is implied along with the Astro Databricks provider

Author: 👀 hmm... oh no I have to pick a side 😅 . I kinda lean towards using Astro Databricks provider, because while it is an implementation way, it is like Cosmos, a provider that is very new and can do pretty amazing things, so whether someone uses this provider or the community-managed Databricks provider does make quite a difference, in my opinion.

description: "Use Airflow, Databricks and the Astro Python SDK in an ELT pipeline to analyze energy data."
id: use-case-airflow-databricks
sidebar_label: "ELT with Airflow + Databricks + Astro SDK"
sidebar_custom_props: { icon: 'img/integrations/databricks.png' }
---

> [Databricks](https://databricks.com/) is a popular unified data and analytics platform built around fully managed Apache Spark clusters. With the [Astro Databricks provider](https://github.com/astronomer/astro-provider-databricks) provider package you can create a Databricks Workflow from existing Databricks notebooks as a task group in your Airflow DAG, allowing you to use Airflow's orchestration capabilities in combination with Databricks' cheapest compute. See also [Why use Airflow with Databricks](airflow-databricks.md#why-use-airflow-with-databricks). The open-source [Astro Python SDK](https://astro-sdk-python.readthedocs.io/en/stable/index.html) greatly simplifies common ELT tasks like loading data and allows users to easily use Pandas on data stored in a data warehouse. 

Reviewer: ```suggestion
[Databricks](https://databricks.com/) is a popular unified data and analytics platform built around fully managed Apache Spark clusters. With the [Astro Databricks provider](https://github.com/astronomer/astro-provider-databricks) package you can create a Databricks Workflow from existing Databricks notebooks as a task group in your Airflow DAG, allowing you to use Airflow's orchestration capabilities in combination with Databricks' cheapest compute. The open-source [Astro Python SDK](https://astro-sdk-python.readthedocs.io/en/stable/index.html) greatly simplifies common ELT tasks like loading data and allows users to easily use Pandas on data stored in a data warehouse. 
```

![DAG graph screenshot](/img/examples/use-case-airflow-databricks_dag_graph_full.png)

After the DAG in this project runs, a graph appears in the `include` directory which shows the combined percentage of solar, hydro and wind energy in a country you selected.

![SHW graph](/img/examples/use-case-airflow-databricks_shw_graph.png)

:::info

For more detailed instructions on using Databricks with the Astro Databricks provider, see the [Databricks tutorial](airflow-databricks.md).
:::

## Before you start

Before trying this example, make sure you have:

- The [Astro CLI](https://docs.astronomer.io/astro/cli/overview).
- [Docker Desktop](https://www.docker.com/products/docker-desktop).
- Access to a Databricks workspace. See [Databricks' documentation](https://docs.databricks.com/getting-started/index.html) for instructions. You can use any workspace that has access to the [Databricks Workflows](https://docs.databricks.com/workflows/index.html) feature. You need a user account with permissions to create notebooks and Databricks jobs. You can use any underlying cloud service, and a [14-day free trial](https://www.databricks.com/try-databricks) is available.

> - Access to an [object storage supported by the Astro Python SDK](https://astro-sdk-python.readthedocs.io/en/stable/supported_file.html). This tutorial uses an [AWS S3](https://aws.amazon.com/s3/) bucket.

Author: github being github, double suggestion :)

- Access to a [relational database supported by the Astro Python SDK](https://astro-sdk-python.readthedocs.io/en/stable/supported_databases.html). This tutorial uses [PostgreSQL](https://www.postgresql.org/).

## Clone the project

> Clone the example project from this [Astronomer GitHub repo](https://github.com/astronomer/learn-airflow-databricks-tutorial). Make sure to create a file called `.env` with the contents of the `.env_example` file in the project root directory and replace the connection details with your own. You will also need to replace the `DATABRICKS_LOGIN_EMAIL`, `S3_BUCKET` and `AWS_REGION` variables with your own values at the start of the DAG code. Change the `COUNTRY` to analyze a different country (for available countries, refer to the source data in the `include` folder).

Reviewer 2: I am probably missing a key detail here, but if the `.env` is already in `.gitignore`, why don't we just keep it there and rename `.env_example` to `.env`? If it's ignored it won't pushed up to GH right?

> Clone the example project from this [Astronomer GitHub repo](https://github.com/astronomer/learn-airflow-databricks-tutorial). Make sure to create a file called `.env` with the contents of the `.env_example` file in the project root directory and replace the connection details with your own. You will also need to replace the `DATABRICKS_LOGIN_EMAIL`, `S3_BUCKET` and `AWS_REGION` variables with your own values at the start of the DAG code. Change the `COUNTRY` to analyze a different country (for available countries, refer to the source data in the `include` folder).

Reviewer 2: Regardless of what we do, we should make it more clear that the reason we're asking users to do this is security related.

## Project contents

### Data source

This example analyzes the share of solar, wind and hydro electricity for different countries. The full source data including other electricity modalities can be found in this [Kaggle dataset](https://www.kaggle.com/datasets/programmerrdai/renewable-energy) derived from [Our World in Data](https://ourworldindata.org/renewable-energy) (License [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)).

The subset of data used in this example can be found in the include folder of the accompanying [GitHub repository](https://github.com/astronomer/learn-airflow-databricks-tutorial/tree/main/include).

### Project code

This project consists of one DAG, [renewable_analysis_dag](https://github.com/astronomer/learn-airflow-databricks-tutorial/blob/main/dags/renewable_analysis_dag.py), which performs an ELT process using the Astro Python SDK and a task group created through the Astro Databricks provider that orchestrates two Databricks notebooks in a Databricks Workflow.

> First, all three separate CSV files, each containing the energy percentages for a list of countries for solar, hydro and wind power respectively are loaded into a new temporary table in the data warehouse using the [Astro Python SDK `load file` operator](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/load_file.html). The operator uses [dynamic task mapping](dynamic-tasks.md) to create a mapped task instance for each file provided to the `input_file` parameter of the `.expand` method, which are then executed in parallel.

Reviewer: ```suggestion
First, three separate CSV files, each containing the energy percentages by country for solar, hydro, and wind power respectively, are loaded into a new temporary table in the data warehouse using the [Astro Python SDK `load file` operator](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/load_file.html). The operator uses [dynamic task mapping](dynamic-tasks.md) to create a mapped task instance for each file provided to the `input_file` parameter of the `.expand` method, which are then executed in parallel.
```

in_tables = aql.LoadFileOperator.partial(
    task_id="in_tables",
    output_table=Table(
        conn_id=DB_CONN_ID,
    ),
).expand(
    input_file=[
        File(path=SOLAR_CSV_PATH),
        File(path=HYDRO_CSV_PATH),
        File(path=WIND_CSV_PATH),
    ]
)
```

Next, the `select_countries` task uses the [Astro Python SDK `@aql.transform` decorator](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/transform.html) to execute a SQL query on each of the three temporary tables created in the `in_tables` task to select only rows for the specified `COUNTRY` and save the results in three new temporary tables. Again, dynamic task mapping is used to execute the SQL queries in parallel on each of the tables contained in the [XCom](airflow-passing-data-between-tasks.md#xcom) output of the previous task (`in_tables.output` provided to `in_table` in `.expand`) in their own dynamically mapped task instances.

```python
@aql.transform
def select_countries(in_table, country):
    return """SELECT * FROM {{ in_table }} WHERE "Entity" = {{ country }};"""

### [...]

country_tables = select_countries.partial(country=COUNTRY).expand(
    in_table=in_tables.output
)
```

> The third task is defined with the [Astro Python SDK ExportToFileOperator](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/export.html) to export the data from the three temporary tables created in the `select_countries` task to an object storage. Dynamic task mapping for this task is achieved using the [`.expand_kwargs` method](dynamic-tasks.md#sets-of-keyword-arguments) since the `input_data` and `output_file` parameter will vary together. The [`.map` method](dynamic-tasks.md#transform-outputs-with-map) is used to modify the output of the upstream task to a list of dictionaries containing the `input_data` and `output_file` parameters for each task instance.

Reviewer 2: ```suggestion
The third task uses the [Astro Python SDK ExportToFileOperator](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/export.html) to export data from the three temporary tables to an object storage. 

The [`.map` method](dynamic-tasks.md#transform-outputs-with-map) converts the output of `country_tables` to a list of dictionaries containing the `input_data` and `output_file` parameters for each task instance. The [`.expand_kwargs` method](dynamic-tasks.md#sets-of-keyword-arguments) then dynamically maps tasks based on each pair of `input_data` and `output_file`. 
```

Author: Thats... so so much better 😅

```python
save_files_to_S3 = aql.ExportToFileOperator.partial(
    task_id="save_files_to_S3",
    if_exists="replace",
).expand_kwargs(
    country_tables.map(

>         lambda x: {

Reviewer 2: Never seen this (`lambda`) before and it doesn't even appear in Airflow docs. Is it worth elaborating on?

Author: Oh! Good point. [Lambda expressions](https://docs.python.org/3/reference/expressions.html?highlight=lambda#lambda) are part of native Python, they are a lesser known way to write functions in one line. 

Basically instead of:

```python
def add_ten(x):
    return x + 10

print(add_ten(2))
```

You can write:

```python
add_ten = lambda x: x + 10
print(add_ten(2))
```

Or even skip naming the function entirely, like I did in `.map`, that is the most common use case, anonymous functions you just need once defined directly where they are needed:

```python
print((lambda x: x + 10)(2))
```

In `.map` `x` will be each element in the XComArg list of `country_tables`. So I transform each table into a dictionary with the input data key being the table and the output file using the name of the table in its path. 

I think I'll add `in a lambda function` or something like that in the sentence above and link the word to the Python docs :)

After these extract steps are completed, a task group created by the [Astro Databricks provider DatabricksWorkflowTaskGroup](https://astronomer.github.io/astro-provider-databricks/#databricks-workflow-taskgroup) completes the transformations:

```python
task_group = DatabricksWorkflowTaskGroup(
    group_id="databricks_workflow",
    databricks_conn_id=DATABRICKS_CONN_ID,
    job_clusters=job_cluster_spec,
)

with task_group:
    notebook_1 = DatabricksNotebookOperator(
        task_id="join_data",
        databricks_conn_id=DATABRICKS_CONN_ID,
        notebook_path=DATABRICKS_NOTEBOOK_PATH_JOIN_DATA,
        source=S3_BUCKET,
        job_cluster_key=DATABRICKS_JOB_CLUSTER_KEY,
    )
    notebook_2 = DatabricksNotebookOperator(
        task_id="transform_data",
        databricks_conn_id=DATABRICKS_CONN_ID,
        notebook_path=DATABRICKS_NOTEBOOK_PATH_TRANSFORM_DATA,
        source=S3_BUCKET,
        job_cluster_key=DATABRICKS_JOB_CLUSTER_KEY,
    )
    notebook_1 >> notebook_2
```

This task group contains three tasks: 

- `launch`: This task is automatically created by the task group and launches a job cluster with the specifications provided in the `job_cluster_spec`.

> - `join_data`: This task executes the first notebook in the Databricks Workflow, which joins the data from the three CSV files created in the object storage into a single CSV file. View the full notebook code [here](https://github.com/astronomer/learn-airflow-databricks-tutorial/blob/main/databricks_notebook_code/join_data_notebook_code.py), note that if you are using an object storage other than S3 you will need to make changes to the code enclosed in `# --------- AWS S3 specific --------- #` comments.

Reviewer 2: ```suggestion
- `join_data`: This task executes the first notebook in the Databricks Workflow, which joins the data from the three CSV files created in the object storage into a single CSV file. View the full notebook code [here](https://github.com/astronomer/learn-airflow-databricks-tutorial/blob/main/databricks_notebook_code/join_data_notebook_code.py). Note that if you are using an object storage other than S3 you will need to make changes to the code enclosed in `# --------- AWS S3 specific --------- #`.
```

> - `transform_data`: This task executes the second notebook in the Databricks Workflow, which fetches the CSV created by the `join_data` task, creates a new column `SHW%` by summing the % points for solar, hydro and wind energy for each year in the specified country and saves the result as a CSV in the object storage. View the full notebook code [here](https://github.com/astronomer/learn-airflow-databricks-tutorial/blob/main/databricks_notebook_code/transform_data_notebook_code.py) (adjust for other object storage solutions as described above).

Reviewer 2: ```suggestion
- `transform_data`: This task executes the second notebook in the Databricks Workflow, which fetches the CSV created by the `join_data` task, creates a new column `SHW%` by summing the % points for solar, hydro, and wind energy for each year, and saves the result as a CSV in the object storage. View the full notebook code [here](https://github.com/astronomer/learn-airflow-databricks-tutorial/blob/main/databricks_notebook_code/transform_data_notebook_code.py). Just like in the previous notebook, you might have to adjust this for other object storage solutions.
```

After the task group is completed, two tasks run in parallel.

> The `delete_intake_files_S3` task cleans up files no longer needed in the S3 bucket using the [S3DeleteObjectsOperator](https://registry.astronomer.io/providers/apache-airflow-providers-amazon/versions/latest/modules/S3DeleteObjectsOperator). If you use a different object storage solution, you will need to define this task using an operator from the corresponding provider. See the [Astronomer Registry](https://registry.astronomer.io/) to explore available providers and operators.

Reviewer 2: ```suggestion
The `delete_intake_files_S3` task uses the [S3DeleteObjectsOperator](https://registry.astronomer.io/providers/apache-airflow-providers-amazon/versions/latest/modules/S3DeleteObjectsOperator) to clean up files which are no longer needed in the S3 bucket. If you use a different object storage solution, you will need to define this task using an operator from the corresponding provider. See the [Astronomer Registry](https://registry.astronomer.io/) to explore available providers and operators.
```

```python
delete_intake_files_S3 = S3DeleteObjectsOperator(
    task_id="delete_intake_files_S3",
    bucket=S3_BUCKET,
    prefix=f"{S3_FOLDER_COUNTRY_SUBSET}/",
    aws_conn_id=AWS_CONN_ID,
)
```

> The the `create_graph` task uses the [Astro Python SDK dataframe operator](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/dataframe.html) to load the information from the combined and transformed CSV created by the Databricks notebook executed via the `transform_data` task into a Pandas dataframe. The task creates a line plot of the % of solar, hydro and wind energy in the specified country over time using [seaborn](https://seaborn.pydata.org/) and [matplotlib](https://matplotlib.org/) and saves it in the `include` folder of the repository.

Reviewer 2: ```suggestion
The `create_graph` task uses the [Astro Python SDK dataframe operator](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/dataframe.html) to load the information from the final CSV into a Pandas dataframe. The task creates a line plot of the % of solar, hydro, and wind energy in the specified country over time using [seaborn](https://seaborn.pydata.org/) and [matplotlib](https://matplotlib.org/), then saves the graph in the `include` folder of the project repository.
```

```python
@aql.dataframe
def create_graph(df: pd.DataFrame):
    sns.set_style("whitegrid")
    sns.lineplot(x="Year", y="SHW%", data=df)
    plt.title(f"% of Solar, Hydro and Wind in the {COUNTRY}")
    plt.xlabel("Year")
    plt.ylabel("Combined SHW (in %)")
    plt.savefig("include/shw.png")

# [...]

load_file_to_db = aql.load_file(
    input_file=File(path=DATABRICKS_RESULT_FILE_PATH, conn_id=AWS_CONN_ID),
    output_table=Table(conn_id=DB_CONN_ID),
)
```

Finally, the [Astro Python SDK `aql.cleanup()`](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/cleanup.html) task removes all temporary tables in the data warehouse that were created by Astro Python SDK tasks.

## See also

> - Tutorial: [Orchestrate dbt Core jobs with Airflow and Cosmos](airflow-databricks.md).

Author: 🤦

- Documentation: [Astro Databricks provider](https://astronomer.github.io/astro-provider-databricks/).

> - Webinar: [The easiest way to orchestrate your dbt workflows from Airflow](https://www.astronomer.io/events/webinars/how-to-orchestrate-databricks-jobs-using-airflow/).

Reviewer: ```suggestion
- Webinar: [How to Orchestrate Databricks Jobs Using Airflow](https://www.astronomer.io/events/webinars/how-to-orchestrate-databricks-jobs-using-airflow/).
```

Author: Leave it to me to somehow change the links but not see the link titles...

This example uses a DAG to extract data from three local CSV files containing the share of solar, hydro and wind electricity in different countries over several years, run a transformation on each file, load the results to S3, and create a line chart of the aggregated data. 

![DAG graph screenshot](/img/examples/use-case-airflow-databricks_dag_graph_full.png)

After the DAG runs, a graph appears in the `include` directory which shows the combined percentage of solar, hydro and wind energy in a country you selected.

![SHW graph](/img/examples/use-case-airflow-databricks_shw_graph.png)

:::info

For more detailed instructions on using Databricks with the Astro Databricks provider, see the [Databricks tutorial](airflow-databricks.md).

:::

## Before you start

Before you try this example, make sure you have:

- The [Astro CLI](https://docs.astronomer.io/astro/cli/overview).
- [Docker Desktop](https://www.docker.com/products/docker-desktop).

> - Access to a Databricks workspace. See [Databricks' documentation](https://docs.databricks.com/getting-started/index.html) for instructions. You can use any workspace that has access to the [Databricks Workflows](https://docs.databricks.com/workflows/index.html) feature. You need a user account with permissions to create notebooks and Databricks jobs. You can use any underlying cloud service, and a [14-day free trial](https://www.databricks.com/try-databricks) is available.

Reviewer 2: ```suggestion
- Access to a Databricks workspace. See [Databricks documentation](https://docs.databricks.com/getting-started/index.html) for instructions. You can use any workspace that has access to the [Databricks Workflows](https://docs.databricks.com/workflows/index.html) feature. You need a user account with permissions to create notebooks and Databricks jobs. You can use any underlying cloud service, and a [14-day free trial](https://www.databricks.com/try-databricks) is available.
```

- Access to an [object storage supported by the Astro Python SDK](https://astro-sdk-python.readthedocs.io/en/stable/supported_file.html). This example uses an [AWS S3](https://aws.amazon.com/s3/) bucket.
- Access to a [relational database supported by the Astro Python SDK](https://astro-sdk-python.readthedocs.io/en/stable/supported_databases.html). This tutorial uses [PostgreSQL](https://www.postgresql.org/).

## Clone the project

> Clone the example project from this [Astronomer GitHub repo](https://github.com/astronomer/learn-airflow-databricks-tutorial). Make sure to create a file called `.env` with the contents of the `.env_example` file in the project root directory and replace the connection details with your own. `.env` is already included in the list of files which git will ignore when creating commits (`.gitignore`). This is necessary to ensure that secrets stored in `.env` are not accidentally pushed to GitHub.

Reviewer 2: ```suggestion
Clone the example project from the [Astronomer GitHub](https://github.com/astronomer/learn-airflow-databricks-tutorial). 

## Edit the project

The project contains a file called `.env_example` which includes placeholder values for your credential details. After you clone the project, create a file called `.env`, copy the contents of `.env_example` into it, and delete `.env_example`. This is required because only `.env` is included in the list of files which Git will ignore when creating commits (`.gitignore`), and cloning an `.env` file directly will cause it to not be ignored the next time you push it to a Git repository.
```

Reviewer 2: I want to make sure that users don't ignore the steps, so I moved the actionable information into an actionable topic. Feel free to override!

## Project contents

### Data source

This example analyzes the share of solar, wind and hydro electricity for different countries. The full source data including other electricity modalities can be found in this [Kaggle dataset](https://www.kaggle.com/datasets/programmerrdai/renewable-energy) derived from [Our World in Data](https://ourworldindata.org/renewable-energy) (License [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)).

The subset of data used in this example can be found in the include folder of the accompanying [GitHub repository](https://github.com/astronomer/learn-airflow-databricks-tutorial/tree/main/include).

### Project code

This project consists of one DAG, [renewable_analysis_dag](https://github.com/astronomer/learn-airflow-databricks-tutorial/blob/main/dags/renewable_analysis_dag.py), which performs an ELT process. It uses the Astro Python SDK and Astro Databricks provider to orchestrate two Databricks notebooks in a Databricks Workflow.

> First, three separate CSV files, each containing the energy percentages by country for solar, hydro, and wind power respectively, are loaded into new tables in the data warehouse using the [Astro Python SDK `load file` operator](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/load_file.html). The operator uses [dynamic task mapping](dynamic-tasks.md) to create a mapped task instance for each dictionary in a list provided to the `.expand_kwargs` method, which are then executed in parallel.

Reviewer 2: ```suggestion
First, three separate CSV files, each containing the energy percentages by country for solar, hydro, and wind power respectively, are loaded into new tables in the data warehouse using the [Astro Python SDK `load file` operator](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/load_file.html). The operator uses [dynamic task mapping](dynamic-tasks.md) to create a mapped task instance for each dictionary in the list provided to the `.expand_kwargs` method. These tasks are then executed in parallel.
```
